controllers/datamovement_controller.go

```diff
@@ -86,314 +83,237 @@ func (r *DataMovementReconciler) Reconcile(ctx context.Context, req ctrl.Request
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
-	// Check if the object is being deleted. Deletion is coordinated around the sub-resources
-	// created or modified as part of data movement.
 	if !dm.GetDeletionTimestamp().IsZero() {
-		log.Info("Starting delete operation")
 
-		if !controllerutil.ContainsFinalizer(dm, finalizer) {
-			return ctrl.Result{}, nil
-		}
+		// TODO: Cancel/Abort the operation from Blake
 
-		teardownFns := []func(context.Context, *nnfv1alpha1.NnfDataMovement) (*ctrl.Result, error){
-			r.teardownRsyncJob,
-			r.teardownLustreJob,
-		}
+		r.completions.Delete(req.Name)
+
+		if controllerutil.ContainsFinalizer(dm, finalizer) {
+			controllerutil.RemoveFinalizer(dm, finalizer)
 
-		for _, teardownFn := range teardownFns {
-			result, err := teardownFn(ctx, dm)
-			log.Info("Teardown", "Result", result, "Error", err)
-			if err != nil {
+			if err := r.Update(ctx, dm); err != nil {
 				return ctrl.Result{}, err
-			} else if result != nil {
-				return *result, nil
 			}
 		}
 
-		controllerutil.RemoveFinalizer(dm, finalizer)
+		return ctrl.Result{}, nil
+	}
+
+	if !controllerutil.ContainsFinalizer(dm, finalizer) {
+		controllerutil.AddFinalizer(dm, finalizer)
 		if err := r.Update(ctx, dm); err != nil {
-			log.Error(err, "Failed to update after removing finalizer")
 			return ctrl.Result{}, err
 		}
 
-		log.Info("Successfully deleted data movement resource")
+		// An update here will cause the reconciler to run again once kubernetes
+		// has recorded the resource it its database.
 		return ctrl.Result{}, nil
 	}
 
-	if !controllerutil.ContainsFinalizer(dm, finalizer) {
+	// Prevent gratuitous wakeups for an resource that is already finished.
+	if dm.Status.State == nnfv1alpha1.DataMovementConditionTypeFinished {
+		return ctrl.Result{}, nil
+	}
 
+	if dm.Status.StartTime.IsZero() {
 		now := metav1.NowMicro()
 		dm.Status.StartTime = &now
+		dm.Status.State = nnfv1alpha1.DataMovementConditionTypeRunning
 
-		// Do first level validation
-		if len(dm.Status.Conditions) == 0 {
-
-			if err := r.validateSpec(dm); err != nil {
-				advanceCondition(dm,
-					nnfv1alpha1.DataMovementConditionTypeFinished,
-					nnfv1alpha1.DataMovementConditionReasonInvalid,
-					fmt.Sprintf("Input validation failed: %v", err),
-				)
-			} else {
-				advanceCondition(dm,
-					nnfv1alpha1.DataMovementConditionTypeStarting,
-					nnfv1alpha1.DataMovementConditionReasonSuccess,
-					"Data movement resource starting",
-				)
-			}
+		if err := r.Status().Update(ctx, dm); err != nil {
+			return ctrl.Result{}, err
+		}
```

Comment:
Line 131 is wrong; it should be moved above the start/initialization. So that leaves the retryOnConflict call once the command completes, which does a Get() first.